workers.py

```diff
@@ -0,0 +1,126 @@
+import logging
+from threading import Thread, current_thread
+
+import requests
+import pandas as pd
+
+import clustering
+# Take the example config for now
+import tests.config as config
+
+
+logger = logging.getLogger()
+MAX_RETRIES = 3
+
+
+def _retryable(method: str, *args, **kwargs) -> requests.Response:
+    """Retryable HTTP request.
+
+    Invoke a "method" on "requests.session" with retry logic.
+    :param method: "get", "post" etc.
+    :param *args: Args for requests (first should be an URL, etc.)
+    :param **kwargs: Kwargs for requests
+    :return: Response object
+    :raises: HTTPError when all requests fail
+    """
+    thread = current_thread()
+
+    with requests.Session() as session:
+        for attempt in range(MAX_RETRIES):
+            try:
+                resp = getattr(session, method)(*args, **kwargs)
+
+                resp.raise_for_status()
+            except (requests.HTTPError, requests.ConnectionError) as e:
+                logger.warning(
+                    '%s: Request failed (attempt #%d), retrying: %s',
+                    thread.name, attempt, str(e)
+                )
+                continue
+            else:
+                return resp
+
+    raise requests.HTTPError('All attempts failed')
+
+
+def _train(data: pd.DataFrame) -> dict:
```

Comment:
For the sake of showing the pipeline it's ok to have the training right here. But later we'd want to separate it out to another container.